core/src/main/java/io/undertow/util/NetworkUtils.java

```diff
@@ -26,9 +26,34 @@
 
 /**
  * @author Stuart Douglas
+ * @author baranowb
  */
 public class NetworkUtils {
 
+    public static final String IP4_EXACT = "(?:\\d{1,3}\\.){3}\\d{1,3}";
+
+    /**
+     * IPV6 match. ?: - unnamed groups are used for performance reasons.
+     * Requirements:
+     * - match full or partial IPV6 ( sliding '::')
+     * - match end to start - ^$ to ensure it does not match part of some random (\d:){n,m}
+     * - IPv4-Embedded IPv6 Address
+     *
+     * NO:
+     * - IPv4 mapped/translated into IPv6
+     *
+     * ^(?:([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}                 - full address
+     * |(?:([0-9a-fA-F]{1,4}:)){1,7}:                               - last compressed
+     * |(?:([0-9a-fA-F]{1,4}:)){1,6}:[0-9a-fA-F]{1,4}               - second to last
+     * |(?:([0-9a-fA-F]{1,4}:)){1,5}(?:(:[0-9a-fA-F]{1,4})){1,2}    - etc
+     * |(?:([0-9a-fA-F]{1,4}:)){1,4}(?:(:[0-9a-fA-F]{1,4})){1,3}
+     * |(?:([0-9a-fA-F]{1,4}:)){1,3}(?:(:[0-9a-fA-F]{1,4})){1,4}
+     * |(?:([0-9a-fA-F]{1,4}:)){1,2}(?:(:[0-9a-fA-F]{1,4})){1,5}
+     * |(?:([0-9a-fA-F])){1,4}:(?:((?:(:[0-9a-fA-F]{1,4})){1,6}))
+     * |:(?:((:[0-9a-fA-F]{1,4}){1,7}|:)))$                         - all the way compressed
+     */
```

Comment:
[Non capturing groups](https://www.regular-expressions.info/brackets.html):
The above expression should be:

either (one non capturing group)

^(
(?:([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4})
|(?:([0-9a-fA-F]{1,4}:){1,7}:)
|(?:([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4})
|(?:([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2})
|(?:([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3})
|(?:([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4})
|(?:([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5})
|(?:[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6}))
|(?::((:[0-9a-fA-F]{1,4}){1,7}|:))
)$

or (multiple non capturing groups)

(
(?:[0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}
|(?:[0-9a-fA-F]{1,4}:){1,7}:
|(?:[0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}
|(?:[0-9a-fA-F]{1,4}:){1,5}(?::[0-9a-fA-F]{1,4}){1,2}
|(?:[0-9a-fA-F]{1,4}:){1,4}(?::[0-9a-fA-F]{1,4}){1,3}
|(?:[0-9a-fA-F]{1,4}:){1,3}(?::[0-9a-fA-F]{1,4}){1,4}
|(?:[0-9a-fA-F]{1,4}:){1,2}(?::[0-9a-fA-F]{1,4}){1,5}
|[0-9a-fA-F]{1,4}:(?:(:[0-9a-fA-F]{1,4}){1,6})
|:(?:?((:[0-9a-fA-F]{1,4}){1,7}|:))
)

in order to be consistent.